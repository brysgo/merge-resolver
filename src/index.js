const flat = require("array.prototype.flat");

const defaultResolver = (values, join) => join(...values);

export const mergeResolver = resolvers => {
  const merge = objects => {
    const combineNext = (path, prevAncestorTypes, ...currentObjects) => {
      let type = resolvers.typeFromObj(
        currentObjects[0],
        path,
        prevAncestorTypes
      );
      const ancestorTypes = [...prevAncestorTypes, type];
      if (!type) {
        if (
          typeof currentObjects[0] === "string" ||
          typeof currentObjects[0] === "number" ||
          typeof currentObjects[0] === "boolean" ||
          typeof currentObjects[0] === "undefined" ||
          currentObjects[0] === null
        )
          return currentObjects[currentObjects.length - 1];
        if (Array.isArray(currentObjects[0])) return flat(currentObjects);
      }

      if (typeof resolvers[type] === "function") {
        // short circuit field level resolution if a function if provided for the type
        return resolvers[type](
          currentObjects,
          combineNext.bind(null, path, ancestorTypes)
        );
      }

      const fieldValues = currentObjects.reduce((acc, cur) => {
        const curType = resolvers.typeFromObj(cur, path, prevAncestorTypes);
        if (curType !== type) {
          throw new Error(`TypeMismatchError: Cannot merge two different types

          Tried to merge "${curType}" with "${type}"`);
        }
        for (let key in cur) {
          acc[key] = acc[key] || [];
          acc[key].push(cur[key]);
        }
        return acc;
      }, {});

      return Object.keys(fieldValues).reduce((acc, cur) => {
        const resolver = (resolvers[type] || {})[cur] || defaultResolver;
        acc[cur] = resolver(
          fieldValues[cur],
          combineNext.bind(null, [...path, cur], ancestorTypes)
        );
        return acc;
      }, {});
    };

    return combineNext([], [], ...objects);
  };
  return merge;
};
