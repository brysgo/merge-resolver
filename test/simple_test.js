import { mergeResolver } from "../src";
import { mergeResolver as mergeResolverDist } from "../build";
import flat from "array.prototype.flat";

flat.shim();

describe("build products", () => {
  it("doesn't break", () => {
    const thing1 = {
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "e" }
        ]
      }
    };
    const thing2 = {
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [{ __typename: "Baz", a: "c", d: "f" }]
      }
    };

    const merge = mergeResolverDist({
      Blah: {
        bazzez: (values, join) => {
          const valuesByKey = {};
          values.flat().forEach(val => {
            valuesByKey[val.a] = valuesByKey[val.a]
              ? join(val, valuesByKey[val.a])
              : val;
          });
          return Object.values(valuesByKey);
        }
      },
      Baz: {
        d: values => values.join("")
      },
      typeFromObj: obj => obj && obj.__typename
    });

    expect(merge([thing1, thing2])).toEqual({
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "fe" }
        ]
      }
    });
  });
});

describe("mergeResolver", () => {
  it("throws an error instead of merging different types", () => {
    const typeA = "A";
    const typeB = "B";
    const merge = mergeResolver({ typeFromObj: o => o });
    expect(() => merge([typeA, typeB])).toThrowErrorMatchingSnapshot();
  });

  it("supports setting types on arrays", () => {
    const merge = mergeResolver({
      Item: ([left, right], join) => {
        return {
          detectedType: "Item",
          ...join(left, right)
        };
      },
      Array: ([left, right], join) => left.map((l, i) => join(l, right[i])),
      typeFromObj: (o, path, ancestorTypes) => {
        if ("Array" === Array.from(ancestorTypes).pop()) {
          return "Item";
        }
        return Array.isArray(o) ? "Array" : undefined;
      }
    });
    expect(
      merge([
        { a: { one: "one" }, b: [{ foo: "bar" }] },
        { a: { one: "two" }, b: [{ baz: "bop" }] }
      ])
    ).toMatchSnapshot();
  });

  it("delegates merge decisions by type", () => {
    const thing1 = {
      __typename: "Query",
      blah: {
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "e" }
        ]
      }
    };
    const thing2 = {
      __typename: "Query",
      blah: {
        bazzez: [{ __typename: "Baz", a: "c", d: "f" }]
      }
    };

    const merge = mergeResolver({
      Blah: {
        bazzez: (values, join) => {
          const valuesByKey = {};
          values.flat().forEach(val => {
            valuesByKey[val.a] = valuesByKey[val.a]
              ? join(val, valuesByKey[val.a])
              : val;
          });
          return Object.values(valuesByKey);
        }
      },
      Baz: {
        d: values => values.join("")
      },
      typeFromObj: (obj, path, ancestorTypes) => {
        if (obj && obj.__typename) return obj.__typename;
        if (
          path &&
          ancestorTypes &&
          Array.from(path).pop() === "blah" &&
          Array.from(ancestorTypes).pop() === "Query"
        )
          return "Blah";
      }
    });

    expect(merge([thing1, thing2])).toEqual({
      __typename: "Query",
      blah: {
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "fe" }
        ]
      }
    });
  });
});

describe("when types are defined without resolvers", () => {
  it("still uses default resolvers", () => {
    const merge = mergeResolver({
      typeFromObj: obj => obj && obj.__typename
    });

    expect(
      merge([
        {
          __typename: "NoResolver",
          a: 1,
          b: { __typename: "And", c: "hello" },
          d: ["ei"]
        },
        {
          __typename: "NoResolver",
          a: 2,
          b: { __typename: "And", c: "world" },
          d: ["eio"]
        }
      ])
    ).toEqual({
      __typename: "NoResolver",
      a: 2,
      b: { __typename: "And", c: "world" },
      d: ["ei", "eio"]
    });
  });
});
