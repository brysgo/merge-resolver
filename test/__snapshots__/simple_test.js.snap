// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeResolver supports setting types on arrays 1`] = `
Object {
  "a": Object {
    "one": "two",
  },
  "b": Array [
    Object {
      "baz": "bop",
      "detectedType": "Item",
      "foo": "bar",
    },
  ],
}
`;

exports[`mergeResolver throws an error instead of merging different types 1`] = `
"TypeMismatchError: Cannot merge two different types

          Tried to merge \\"B\\" with \\"A\\""
`;
